### PR TITLE
Added a specific dynamic list component - EUI-3045

### PR DESCRIPTION
**Pull request by phillip-whitaker**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/EUI-3045

### Change description ###

Created a specific dynamic read/write dropdown field. The new write dropdown component checks whether a value property has already been set from the API and if not, sets the value to null so that the default "-- Select an option --" is shown rather than a blank dropdown 

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[X] No
```

### CODE REVIEW

File: src/shared/components/palette/fixed-list/read-fixed-list-field.component.ts [hunk 1]
```diff
@@ -1,20 +1,10 @@
...
-    /**
-     *
-     * Reassigning list_items from formatted_list when list_items is empty for DynamicList's
-     */
-    if (!this.caseField.list_items && this.caseField.formatted_value && this.caseField.formatted_value.list_items) {
-      this.caseField.list_items = this.caseField.formatted_value.list_items;
-    }
-
-  }
 }
```
Comment (phillip-whitaker) [inline src/shared/components/palette/fixed-list/read-fixed-list-field.component.ts:10]: This has been removed as the dynamic list now is in its own component so does not need this clause in here

File: src/shared/components/palette/fixed-list/write-fixed-list-field.component.ts [hunk 2]
```diff
@@ -11,19 +11,12 @@ export class WriteFixedListFieldComponent extends AbstractFieldWriteComponent im
...
   ngOnInit() {
-
-    /**
-     *
-     * Reassigning list_items from formatted_list when list_items is empty for DynamicList's
-     */
-    if (!this.caseField.list_items && this.caseField.formatted_value && this.caseField.formatted_value.list_items) {
-      this.caseField.list_items = this.caseField.formatted_value.list_items;
-    }
-
```
Comment (phillip-whitaker) [inline src/shared/components/palette/fixed-list/write-fixed-list-field.component.ts:22]: This has been removed as the dynamic list now is in its own component so does not need this clause in here

File: src/shared/components/palette/fixed-list/write-fixed-list-field.component.ts [hunk 3]
```diff
@@ -11,19 +11,12 @@ export class WriteFixedListFieldComponent extends AbstractFieldWriteComponent im
...
     let isNull = this.caseField.value === undefined || this.caseField.value === '';
```
Comment (paulhowes-HMCTS) [inline src/shared/components/palette/fixed-list/write-fixed-list-field.component.ts:14]: Could use dual NOT operators to reduce this code, i.e. if (!!this.caseField.value)

File: src/shared/components/palette/dynamic-list/write-dynamic-list-field.html [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+    </label>
+  
+    <select class="form-control ccd-dropdown bottom-30" [id]="id()" 
+    [(ngModel)]="caseField.value" [formControl]="dynamicListFormControl">
+      <option [ngValue]=null>--Select a value--</option>
+      <option [ngValue]="type.code" *ngFor="let type of caseField.list_items">{{type.label}}</option>
+    </select>
+  
+  </div>
+
```
Comment (paulhowes-HMCTS) [inline src/shared/components/palette/dynamic-list/write-dynamic-list-field.html:16]: Getting a warning about no newline at end of file here...

File: src/shared/components/palette/dynamic-list/write-dynamic-list-field.html [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+      <span class="form-label" *ngIf="caseField.label">{{caseField | ccdFieldLabel}}</span>
+      <span class="form-hint" *ngIf="caseField.hint_text">{{caseField.hint_text}}</span>
+      <span class="error-message" *ngIf="dynamicListFormControl.errors && dynamicListFormControl.touched">{{dynamicListFormControl.errors | ccdFirstError}}</span>
+    </label>
+  
+    <select class="form-control ccd-dropdown bottom-30" [id]="id()" 
+    [(ngModel)]="caseField.value" [formControl]="dynamicListFormControl">
+      <option [ngValue]=null>--Select a value--</option>
+      <option [ngValue]="type.code" *ngFor="let type of caseField.list_items">{{type.label}}</option>
+    </select>
```
Comment (paulhowes-HMCTS) [inline src/shared/components/palette/dynamic-list/write-dynamic-list-field.html:13]: Could we end up with an empty label if none of the *ngIf cases are true? Suggest adding all ngIf's to label

Comment (paulhowes-HMCTS) [reply src/shared/components/palette/dynamic-list/write-dynamic-list-field.html:13]: Also would the select and spans conform to accessibility rules?

File: src/shared/components/palette/dynamic-list/write-dynamic-list-field.component.ts [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+  ngOnInit() {
+    /**
+     *
+     * Reassigning list_items from formatted_list when list_items is empty
+     */
+    if (!this.caseField.list_items && this.caseField.formatted_value && this.caseField.formatted_value.list_items) {
+      this.caseField.list_items = this.caseField.formatted_value.list_items;
+    }
+
+    let isNull = this.caseField.value === undefined || this.caseField.value === '';
```
Comment (paulhowes-HMCTS) [inline src/shared/components/palette/dynamic-list/write-dynamic-list-field.component.ts:22]: Could use dual NOT operators to reduce this code, i.e. if (!!this.caseField.value)

**Review by paulhowes-HMCTS — CHANGES_REQUESTED** [review]

File: src/shared/components/palette/fixed-list/write-fixed-list-field.component.ts (hunk 3 above)
Comment (phillip-whitaker) [reply src/shared/components/palette/fixed-list/write-fixed-list-field.component.ts:14]: done

File: src/shared/components/palette/dynamic-list/write-dynamic-list-field.html (hunk 4 above)
Comment (phillip-whitaker) [reply src/shared/components/palette/dynamic-list/write-dynamic-list-field.html]: done

File: src/shared/components/palette/dynamic-list/write-dynamic-list-field.html (hunk 5 above)
Comment (phillip-whitaker) [reply src/shared/components/palette/dynamic-list/write-dynamic-list-field.html]: Accessibility I have added a role attrribute to the <option> DOM element. and as a backup I have added text in the label element if there is no label and no hint text - ensuring there will always be something shown.

File: src/shared/components/palette/dynamic-list/write-dynamic-list-field.component.ts (hunk 6 above)
Comment (phillip-whitaker) [reply src/shared/components/palette/dynamic-list/write-dynamic-list-field.component.ts]: done

**Review by paulhowes-HMCTS — APPROVED** [review]

**Comment by phillip-whitaker** [comment]
This PR has been superseded by: https://github.com/hmcts/ccd-case-ui-toolkit/pull/684

**Pull request closed without merging**